src/common/linebuf-reader.cc

```diff
@@ -49,11 +50,25 @@ const char* LinebufReader::IncompleteLine() {
 
 const char* LinebufReader::ReadLine() {
   for (char *i = content_start_; i < content_end_; ++i) {
-    if (*i == '\n' || *i == '\r') {
-      *i = '\0';
-      const char *line = content_start_;
-      content_start_ = i + 1;
-      return line;
+    switch (ch_state_) {
+      case LF_SEEN: {
+        // Silently ignore newline
+        if (*i == '\n') {
+          ch_state_ = CR_SEEN;
+          content_start_ = i + 1;
+          continue;
+        }
+      }
+      case CR_SEEN: case IN_LINE: {
```

Comment:
For readability, I'd put the different cases under each other:
```
case CR_SEEN:  /* fallthrough */
case IN_LINE:  {
   // ...
}
```

Anyway, looking at this, it seems my suggestion of three states was not needed: one of these is redundant. We only really need `IN_LINE` and `CR_SEEN`.

State transitions  @(`{START_STATE} -- {character} --> [{optional action} {NEW_STATE}`)
```
IN_LINE -- \n --> [emitting line] IN_LINE
IN_LINE -- \r --> [emitting line] CR_SEEN
IN_LINE --(any other char) --> IN_LINE
CR_SEEN -- \n --> [skip character] --> IN_LINE
CR_SEEN -- \r --> [emit empty line] --> CR_SEEN
CR_SEEN --(any other char) --> IN_LINE;
```

Reply:
(solved) yes I saw that the only important state is the CR_SEEN so I converted the mini fsm into a single bool. I was going to work on that but you anticipated me :smiley: